File: frontend/src/types/events.ts
/**
 * Event types for the Lite Agent frontend.
 * Based on the Agent User Interaction Protocol (AG-UI).
 */

// ============================================================================
// Base Event
// ============================================================================

export interface BaseEvent {
    type: string;
    timestamp: string;
    raw_event?: any;
}

// ============================================================================
// Lifecycle Events
// ============================================================================

export interface RunStarted extends BaseEvent {
    type: 'RunStarted';
    run_id: string;
    session_id?: string;
}

export interface RunFinished extends BaseEvent {
    type: 'RunFinished';
    run_id: string;
    duration_ms?: number;
    total_cost_usd?: number;
    usage?: {
        input_tokens?: number;
        output_tokens?: number;
        cache_creation_input_tokens?: number;
        cache_read_input_tokens?: number;
        total_cost_usd?: number;
        [key: string]: any;
    };
}

export interface RunError extends BaseEvent {
    type: 'RunError';
    run_id: string;
    error: string;
    error_type?: string;
}

export interface StepStarted extends BaseEvent {
    type: 'StepStarted';
    step_id: string;
    run_id: string;
    step_name?: string;
}

export interface StepFinished extends BaseEvent {
    type: 'StepFinished';
    step_id: string;
    run_id: string;
    duration_ms?: number;
}

// ============================================================================
// Text Message Events
// ============================================================================

export interface TextMessageStart extends BaseEvent {
    type: 'TextMessageStart';
    message_id: string;
    role: 'user' | 'assistant' | 'system';
}

export interface TextMessageContent extends BaseEvent {
    type: 'TextMessageContent';
    message_id: string;
    delta: string;
}

export interface TextMessageEnd extends BaseEvent {
    type: 'TextMessageEnd';
    message_id: string;
}

// ============================================================================
// Tool Call Events
// ============================================================================

export interface ToolCallStart extends BaseEvent {
    type: 'ToolCallStart';
    tool_call_id: string;
    tool_call_name: string;
    parent_message_id?: string;
}

export interface ToolCallArgs extends BaseEvent {
    type: 'ToolCallArgs';
    tool_call_id: string;
    delta: string;
}

export interface ToolCallEnd extends BaseEvent {
    type: 'ToolCallEnd';
    tool_call_id: string;
}

export interface ToolCallResult extends BaseEvent {
    type: 'ToolCallResult';
    message_id: string;
    tool_call_id: string;
    content: any;
    role: 'tool';
    is_error: boolean;
    metadata?: {
        content_type?: string; // 'image' | 'video' | 'audio' | 'url' | 'json' | 'text' | 'file' | 'list' | 'unknown'
        media_type?: string; // MIME type (e.g., 'image/jpeg', 'image/png')
        url?: string; // URL for remote resources
        data?: any; // Additional data (base64, JSON, etc.)
        encoding?: string; // 'base64' for base64-encoded content
        file_path?: string; // File path for local files
        source_key?: string; // Key in JSON that contains the content
        items?: any[]; // For list types
    };
}

// ============================================================================
// Thinking/Reasoning Events
// ============================================================================

export interface ThinkingStart extends BaseEvent {
    type: 'ThinkingStart';
    thinking_id: string;
    parent_message_id?: string;
}

export interface ThinkingContent extends BaseEvent {
    type: 'ThinkingContent';
    thinking_id: string;
    delta: string;
}

export interface ThinkingEnd extends BaseEvent {
    type: 'ThinkingEnd';
    thinking_id: string;
}

// ============================================================================
// State Management Events
// ============================================================================

export interface StateSnapshot extends BaseEvent {
    type: 'StateSnapshot';
    state: Record<string, any>;
}

export interface StateDelta extends BaseEvent {
    type: 'StateDelta';
    delta: Record<string, any>;
}

// ============================================================================
// Generative UI Events (AG-UI Protocol)
// ============================================================================

export interface UIComponent extends BaseEvent {
    type: 'UIComponent';
    component_id: string;
    component_type: string; // e.g., 'button', 'image', 'form', 'card', 'chart', etc.
    props: Record<string, any>; // Component properties
    children?: Array<Record<string, any>>; // Nested components
    constraints?: Record<string, any>; // Validation constraints
    parent_component_id?: string; // For nested components
    message_id?: string; // Associated message
}

export interface UIUpdate extends BaseEvent {
    type: 'UIUpdate';
    component_id: string;
    props: Record<string, any>; // Updated properties
    merge?: boolean; // Whether to merge with existing props
}

export interface UIRemove extends BaseEvent {
    type: 'UIRemove';
    component_id: string;
}

export interface UIInteraction extends BaseEvent {
    type: 'UIInteraction';
    component_id: string;
    interaction_type: string; // e.g., 'click', 'submit', 'change', 'focus', etc.
    data: Record<string, any>; // Interaction data
}

// ============================================================================
// Custom/Extension Events
// ============================================================================

export interface AgentCustomEvent extends BaseEvent {
    type: string;
    data: Record<string, any>;
}

// ============================================================================
// Event Union Type
// ============================================================================

export type AgentEvent =
    | RunStarted
    | RunFinished
    | RunError
    | StepStarted
    | StepFinished
    | TextMessageStart
    | TextMessageContent
    | TextMessageEnd
    | ToolCallStart
    | ToolCallArgs
    | ToolCallEnd
    | ToolCallResult
    | ThinkingStart
    | ThinkingContent
    | ThinkingEnd
    | StateSnapshot
    | StateDelta
    | UIComponent
    | UIUpdate
    | UIRemove
    | UIInteraction
    | AgentCustomEvent;

// ============================================================================
// Event Handler Types
// ============================================================================

export type EventHandler<T extends AgentEvent = AgentEvent> = (event: T) => void;

export interface EventHandlers {
    onRunStarted?: EventHandler<RunStarted>;
    onRunFinished?: EventHandler<RunFinished>;
    onRunError?: EventHandler<RunError>;
    onStepStarted?: EventHandler<StepStarted>;
    onStepFinished?: EventHandler<StepFinished>;
    onTextMessageStart?: EventHandler<TextMessageStart>;
    onTextMessageContent?: EventHandler<TextMessageContent>;
    onTextMessageEnd?: EventHandler<TextMessageEnd>;
    onToolCallStart?: EventHandler<ToolCallStart>;
    onToolCallArgs?: EventHandler<ToolCallArgs>;
    onToolCallEnd?: EventHandler<ToolCallEnd>;
    onToolCallResult?: EventHandler<ToolCallResult>;
    onThinkingStart?: EventHandler<ThinkingStart>;
    onThinkingContent?: EventHandler<ThinkingContent>;
    onThinkingEnd?: EventHandler<ThinkingEnd>;
    onStateSnapshot?: EventHandler<StateSnapshot>;
    onStateDelta?: EventHandler<StateDelta>;
    onUIComponent?: EventHandler<UIComponent>;
    onUIUpdate?: EventHandler<UIUpdate>;
    onUIRemove?: EventHandler<UIRemove>;
    onUIInteraction?: EventHandler<UIInteraction>;
    onCustomEvent?: EventHandler<AgentCustomEvent>;
    onAnyEvent?: EventHandler<AgentEvent>;
}

File: backend/core/events.py
"""
Event types for the Lite Agent system.
Based on the Agent User Interaction Protocol (AG-UI).
"""

from typing import Literal, Optional, Any, Dict
from pydantic import BaseModel, Field, ConfigDict
from datetime import datetime


# ============================================================================
# Base Event
# ============================================================================

class BaseEvent(BaseModel):
    """Base class for all events."""
    model_config = ConfigDict(populate_by_name=True)  # Use snake_case for all fields
    
    type: str
    timestamp: datetime = Field(default_factory=datetime.now)
    raw_event: Optional[Dict[str, Any]] = None


# ============================================================================
# Lifecycle Events
# ============================================================================

class RunStarted(BaseEvent):
    """Emitted when an agent run starts."""
    type: Literal["RunStarted"] = "RunStarted"
    run_id: str
    session_id: Optional[str] = None


class RunFinished(BaseEvent):
    """Emitted when an agent run completes successfully."""
    type: Literal["RunFinished"] = "RunFinished"
    run_id: str
    duration_ms: Optional[int] = None
    total_cost_usd: Optional[float] = None
    usage: Optional[Dict[str, Any]] = None  # Token usage details


class RunError(BaseEvent):
    """Emitted when an agent run encounters an error."""
    type: Literal["RunError"] = "RunError"
    run_id: str
    error: str
    error_type: Optional[str] = None


class StepStarted(BaseEvent):
    """Emitted when a step within a run starts."""
    type: Literal["StepStarted"] = "StepStarted"
    step_id: str
    run_id: str
    step_name: Optional[str] = None


class StepFinished(BaseEvent):
    """Emitted when a step completes."""
    type: Literal["StepFinished"] = "StepFinished"
    step_id: str
    run_id: str
    duration_ms: Optional[int] = None


# ============================================================================
# Text Message Events
# ============================================================================

class TextMessageStart(BaseEvent):
    """Emitted when a text message begins."""
    type: Literal["TextMessageStart"] = "TextMessageStart"
    message_id: str
    role: Literal["user", "assistant", "system"] = "assistant"


class TextMessageContent(BaseEvent):
    """Emitted for each chunk of text content."""
    type: Literal["TextMessageContent"] = "TextMessageContent"
    message_id: str
    delta: str  # The incremental text content


class TextMessageEnd(BaseEvent):
    """Emitted when a text message completes."""
    type: Literal["TextMessageEnd"] = "TextMessageEnd"
    message_id: str


# ============================================================================
# Tool Call Events
# ============================================================================

class ToolCallStart(BaseEvent):
    """Emitted when a tool call begins."""
    type: Literal["ToolCallStart"] = "ToolCallStart"
    tool_call_id: str
    tool_call_name: str
    parent_message_id: Optional[str] = None


class ToolCallArgs(BaseEvent):
    """Emitted for streaming tool arguments."""
    type: Literal["ToolCallArgs"] = "ToolCallArgs"
    tool_call_id: str
    delta: str  # Incremental JSON arguments


class ToolCallEnd(BaseEvent):
    """Emitted when a tool call completes."""
    type: Literal["ToolCallEnd"] = "ToolCallEnd"
    tool_call_id: str


class ToolCallResult(BaseEvent):
    """Emitted when a tool returns a result."""
    type: Literal["ToolCallResult"] = "ToolCallResult"
    message_id: str
    tool_call_id: str
    content: Any
    role: Literal["tool"] = "tool"
    is_error: bool = False
    metadata: Optional[Dict[str, Any]] = None  # Content type metadata for dynamic UI rendering


# ============================================================================
# Thinking/Reasoning Events
# ============================================================================

class ThinkingStart(BaseEvent):
    """Emitted when thinking/reasoning begins."""
    type: Literal["ThinkingStart"] = "ThinkingStart"
    thinking_id: str
    parent_message_id: Optional[str] = None


class ThinkingContent(BaseEvent):
    """Emitted for streaming thinking content."""
    type: Literal["ThinkingContent"] = "ThinkingContent"
    thinking_id: str
    delta: str


class ThinkingEnd(BaseEvent):
    """Emitted when thinking completes."""
    type: Literal["ThinkingEnd"] = "ThinkingEnd"
    thinking_id: str


# ============================================================================
# State Management Events
# ============================================================================

class StateSnapshot(BaseEvent):
    """Complete state snapshot."""
    type: Literal["StateSnapshot"] = "StateSnapshot"
    state: Dict[str, Any]


class StateDelta(BaseEvent):
    """Incremental state change."""
    type: Literal["StateDelta"] = "StateDelta"
    delta: Dict[str, Any]


# ============================================================================
# Generative UI Events (AG-UI Protocol)
# ============================================================================

class UIComponent(BaseEvent):
    """
    Generative UI component event.
    Allows agents to dynamically generate UI components.
    Based on AG-UI Generative UI specification.
    """
    type: Literal["UIComponent"] = "UIComponent"
    component_id: str
    component_type: str  # e.g., 'button', 'image', 'form', 'card', 'chart', etc.
    props: Dict[str, Any]  # Component properties
    children: Optional[list[Dict[str, Any]]] = None  # Nested components
    constraints: Optional[Dict[str, Any]] = None  # Validation constraints
    parent_component_id: Optional[str] = None  # For nested components
    message_id: Optional[str] = None  # Associated message


class UIUpdate(BaseEvent):
    """Update an existing UI component."""
    type: Literal["UIUpdate"] = "UIUpdate"
    component_id: str
    props: Dict[str, Any]  # Updated properties
    merge: bool = True  # Whether to merge with existing props


class UIRemove(BaseEvent):
    """Remove a UI component."""
    type: Literal["UIRemove"] = "UIRemove"
    component_id: str


class UIInteraction(BaseEvent):
    """
    User interaction with a UI component.
    This is sent from frontend to backend when user interacts with generated UI.
    """
    type: Literal["UIInteraction"] = "UIInteraction"
    component_id: str
    interaction_type: str  # e.g., 'click', 'submit', 'change', 'focus', etc.
    data: Dict[str, Any]  # Interaction data
    timestamp: datetime = Field(default_factory=datetime.now)


# ============================================================================
# Custom/Extension Events
# ============================================================================

class CustomEvent(BaseEvent):
    """Custom event for extensions."""
    type: str  # Custom type name
    data: Dict[str, Any]


# ============================================================================
# Event Union Type
# ============================================================================

AgentEvent = (
    RunStarted | RunFinished | RunError |
    StepStarted | StepFinished |
    TextMessageStart | TextMessageContent | TextMessageEnd |
    ToolCallStart | ToolCallArgs | ToolCallEnd | ToolCallResult |
    ThinkingStart | ThinkingContent | ThinkingEnd |
    StateSnapshot | StateDelta |
    UIComponent | UIUpdate | UIRemove | UIInteraction |
    CustomEvent
)
